src/diffTool/jsonDiff.ts

```diff
@@ -0,0 +1,374 @@
+  Object.keys(graphDiffs).forEach(key => {
+    if (key === JSON_DIFF_PATCH_ARRAY_KEY) {
+      //Ignoring the property added by jsondiffpath to indicate that graph is an array
+      return;
+    }
+    const diff = graphDiffs[key];
+    let typedDiff;
+    if (_.isArray(diff)) {
+      //jsondiffpatch returns each diff as an array. So an array in place of node indicates that array is added/removed/moved
+      typedDiff = addNodeDiff(diff);
+    } else if (_.isObject(diff)) {
+      //properties in a node are added/removed/modified/moved
+      typedDiff = parseNodePropDiffs(diff[JSON_LD_KEY_ID], diff);
+    }
+    if (typedDiff != null) {
+      typedDiffs.push(typedDiff);
+    }
+  });
+  return typedDiffs;
+}
+
+/**
+ * Parse differences to the node properties
+ * @param nodeId ID of the node
+ * @param diff JSON node that has differences
+ *
+ * @returns Instance of NodeDiff for the node if there are differences
+ */
+function parseNodePropDiffs(nodeId: string, diff: object): NodeDiff {
+  let typedDiff = new NodeDiff(nodeId);
+  Object.keys(diff)
+    .filter(key => key !== JSON_LD_KEY_ID) //ignore node id
+    .forEach(key => {
+      const value = diff[key];
+      /**
+       * check if the property of the node is an array. Key is the index of the original array
+       */
+      if (value[JSON_DIFF_PATCH_ARRAY_KEY] === JSON_DIFF_PATCH_ARRAY_VALUE) {
+        //Ignore property added by jsondiffpath to indicate array property
+        Object.keys(value)
+          .filter(arrKey => arrKey !== JSON_DIFF_PATCH_ARRAY_KEY)
+          .forEach(arrKey => {
+            //Since this is flattened graph array value is either a string or a object with an id referencing other object in the graph
+            const arrValue = value[arrKey];
+            const diffType = getDiffType(arrValue);
+            ramlToolLogger.debug(
+              `Adding difference of node: ${typedDiff.id}, array property: ${key}, array key: ${arrKey}, difference type: ${DiffType[diffType]}`
+            );
+            addNodeArrayPropertyDiffs(key, arrValue, diffType, typedDiff);
+          });
+      } else {
+        const diffType = getDiffType(value);
+        ramlToolLogger.debug(
+          `Adding difference of node: ${typedDiff.id}, property: ${key}, difference type: ${DiffType[diffType]}`
+        );
+        addNodePropertyDiffs(key, value, diffType, typedDiff);
+      }
+    });
+  if (_.isEmpty(typedDiff.added) && _.isEmpty(typedDiff.removed)) {
+    //when there are just moves and no actual changes, ignore the typed diff
+    typedDiff = undefined;
+  }
+  return typedDiff;
+}
+
+/**
+ * Get the type of the difference from the difference array generated by jsondiffpatch
+ * @param diff Difference of a node or a node property
+ *
+ * @return DiffType Type of the difference
+ */
+function getDiffType(diff: any[]): DiffType {
```

Comment:
We should try to avoid explicit `any`. (There's even a lint rule for this.) Based on the test files, a more accurate type would be `(object | string | number)[]`. Even better would be enumerating all combinations, if possible and not excessively verbose. Because this is used in multiple places, I'd recommend declaring a type, maybe `DiffSignature` or something?

Reply:
good suggestion but this is really any at least for now. I don't want to guess the types that graph can have